src/zarr/core/array.py

```diff
@@ -677,28 +686,40 @@ def _create_metadata_v3(
         """
 
         shape = parse_shapelike(shape)
-        codecs = list(codecs) if codecs is not None else _get_default_codecs(np.dtype(dtype))
+        if codecs is None:
+            filters, serializer, compressors = _get_default_chunk_encoding_v3(dtype)
+            codecs_parsed = (*filters, serializer, *compressors)
+        else:
+            codecs_parsed = tuple(codecs)
+
         chunk_key_encoding_parsed: ChunkKeyEncodingLike
         if chunk_key_encoding is None:
             chunk_key_encoding_parsed = {"name": "default", "separator": "/"}
         else:
             chunk_key_encoding_parsed = chunk_key_encoding
 
-        if dtype.kind in "UTS":
+        if dtype.to_dtype().kind in ("U", "T", "S"):
             warn(
                 f"The dtype `{dtype}` is currently not part in the Zarr format 3 specification. It "
                 "may not be supported by other zarr implementations and may change in the future.",
                 category=UserWarning,
                 stacklevel=2,
             )
+
+        if fill_value is None:
+            # v3 spec will not allow a null fill value
+            fill_value_parsed = dtype.default_value()
+        else:
+            fill_value_parsed = fill_value
```

Comment:
should (or are) we be casting this scalar somewhere?

Reply:
Good point, we can use `dtype.cast_value()` here.